File: soln-array-problems/4sum.java
/**
 @author Farheen Bano
  
 References-
 https://www.interviewbit.com/problems/4-sum/
 https://leetcode.com/problems/4sum
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Set<List<Integer>> result=new HashSet<>();
        if(nums.length<4)
            return new ArrayList<>(result);
        
        Arrays.sort(nums);
        for(int i=0;i<nums.length;i++){
            for(int j=i+1;j<nums.length;j++){
                int k=j+1;
                int l=nums.length-1;
                while(k<l){
                    int sum=nums[i]+nums[j]+nums[k]+nums[l];
                    if(sum==target){
                        List<Integer> list=Arrays.asList(nums[i],nums[j],nums[k],nums[l]);
                        result.add(list);       
                    }
                    if(sum<=target)
                        k++;
                    else
                        l--;                        
                }
            }
        }
        return new ArrayList<>(result);
    }
}

File: C++/soln-array-problems/4sum.cpp
/**
 @author Farheen Bano
  
 Date 15-07-2021

 References-
 https://www.interviewbit.com/problems/4-sum/
 https://leetcode.com/problems/4sum
*/

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> output_quad;
        
        int len=nums.size();
        int left,right,sum;
        sort(nums.begin(),nums.end());
        
        for(int i=0;i<len-3;i++){
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }
            
            for(int j=i+1;j<len-2;j++){
                if(j>i+1 && nums[j]==nums[j-1])
                    continue;
                
                left=j+1, right=len-1;
                while(left<right){
                    sum=nums[i]+nums[j]+nums[left]+nums[right];
                    if(target==sum){
                        output_quad.push_back({nums[i],nums[j],nums[left],nums[right]});
                    
                        left++;
                        right--;

                        while(left<right && nums[right]==nums[right+1])
                            right--;
                        while(left<right && nums[left]==nums[left-1])
                            left++;
                    }
                    else if(sum>target){
                        right--;
                    } 
                    else
                        left++;
                }
            }
        }
        return output_quad;
    }
};